### ISSUE TITLE
Cache cannot be set to no in /etc/systemd/resolved.conf

### ISSUE THREAD

**Issue by robertvargason** [camptocamp/puppet-systemd#113]
As far as I understand one should be able to set the value of Cache to 'yes' or 'no' in /etc/systemd/resolved.conf via systemd::resolved.

But in systemd::resolved the variable named $cache is a Boolean and if it is set to false, the value does not get rendered to /etc/systemd/resolved.conf:

```
$_cache = $cache ? {
    true    => 'yes',
    false   => 'no',
  }

  if $cache {
    ini_setting{ 'cache':
      ensure  => 'present',
      value   => $_cache,
      setting => 'Cache',
      section => 'Resolve',
      path    => '/etc/systemd/resolved.conf',
      notify  => Service['systemd-resolved'],
    }
  }
```

So only setting it to true will render it to the file with Cache=yes (which under Ubuntu is the default anyway) and one can never set Cache=no. Because using only a Boolean will mean you have to render it out everywhere (even on installations where you are not concerned with this value and would like not to have this file modified), I would hope for a different solution (e.g. not using a Boolean or two variables).

Thank you so much!

**Comment by raphink** [comment]
I can confirm this bug, as the code reads:

```puppet
  $_cache = $cache ? {                                                          
    true    => 'yes',                                                           
    false   => 'no',                                                            
  }                                                                             
                                                                                
  if $cache {                                                                   
    ini_setting{ 'cache':                                                       
      ensure  => 'present',                                                     
      value   => $_cache,                                                       
      setting => 'Cache',                                                       
      section => 'Resolve',                                                     
      path    => '/etc/systemd/resolved.conf',                                  
      notify  => Service['systemd-resolved'],                                   
    }                                                                           
  }   
```

so `$_cache` gets set to the right value, but never applied.

**Comment by Schamane187** [comment]
Hi,

just ran into an issue as well, I want to use the parameter no-negative in section cache.
I had to change manifest to enable caching with 
"no-negative"

so in
resolved.pp

 ` Boolean $cache                                                   = $systemd::cache,`
has been changed to  
`  Optional[Variant[Array[String],String]] $cache                   = $systemd::cache,`

and in section 

 ```
 $_cache = $cache ? {
    true          => 'yes',
    false         => 'no',
    'no-negative' => 'no-negative',
  }

```
no-negative was added

Could u update or built in using the Feature no-negative in a future release? 
Thx in advance

**Comment by bastelfreak** [comment]
@Schamane187 can you provide a patch for this?

**Comment by Schamane187** [comment]
Hi,

was moving away for the moment from systemd for that task, cause CentOS 7 Systemd is to outdated for those features. But with CentOS I will be back to this.

A Patch for  2 lines? but ok, here we go 

```
diff -upr puppet-systemd-master_org/manifests/resolved.pp puppet-systemd-master/manifests/resolved.pp
--- puppet-systemd-master_org/manifests/resolved.pp	2019-07-15 11:48:00.000000000 +0200
+++ puppet-systemd-master/manifests/resolved.pp	2019-09-23 09:55:40.000000000 +0200
@@ -53,8 +53,8 @@ class systemd::resolved (
   Optional[Variant[Boolean,Enum['resolve']]] $multicast_dns          = $systemd::multicast_dns,
   Optional[Variant[Boolean,Enum['allow-downgrade']]] $dnssec         = $systemd::dnssec,
   Optional[Variant[Boolean,Enum['opportunistic', 'no']]] $dnsovertls = $systemd::dnsovertls,
-  Boolean $cache                                                     = $systemd::cache,
   Optional[Variant[Boolean,Enum['udp', 'tcp']]] $dns_stub_listener   = $systemd::dns_stub_listener,
+  Optional[Variant[Array[String],String]] $cache 		     = $systemd::cache,
   Boolean $use_stub_resolver                                         = $systemd::use_stub_resolver,
 ){

@@ -200,6 +200,7 @@ class systemd::resolved (
   $_cache = $cache ? {
     true    => 'yes',
     false   => 'no',
+   'no-negative' => 'no-negative',
   }

   if $cache {
```
Cheers

**Comment by schlitzered** [comment]
this pull request solves the issue:

https://github.com/camptocamp/puppet-systemd/pull/140

**Comment by bastelfreak** [comment]
PR got merged

**Comment by robertvargason** [comment]
This issue should not be closed as the original issue is still not resolved. One cannot set Cache=no

**Comment by schlitzered** [comment]
@robertvargason: you are right. there was an "if $cache" block, which prevent setting this to false.

i created a new merge request for this.

**Comment by robertvargason** [comment]
@schlitzered: thanks for that. the commited version is working for us.

if you want to be very clever with your implementation you might want to consider one of the two extra solutions:

a) to introduce a setting which leaves the setting Cache unrendered. So that Cache=yes, Cache=no, Cache=no-negative as well as no Cache setting at all (should be the module default) is possible. which leaves the option to not render the value and let the default value for Cache from systemd rule supreme (which is at the moment 'yes'). meaning also that the new puppet module version does not touch the /etc/systemd/resolved.conf (which is nice in my opinion).

b) use 'yes' as the option to not render the setting at all, because it is the systemd default. Meaning only 'no' or 'no-negative' is rendered. There is no benefit to explicit set Cache=yes, because it is the systemd default. This would (if you make 'yes' the module default value) also leave resolved.conf files untouched (except when module users already used to set 'yes' [which was unnecessary]).

thanks a lot

**Comment by schlitzered** [comment]
i leave this up to the original maintainer to decide. 

but IMHO, I would set the default in the puppet module, to whatever the default is for the $OS_NAME and $OS_MAJOR. and always put this config into the on the system.

**Comment by raphink** [comment]
FIxed in #140
